### ISSUE TITLE
Firebase hosting

### ISSUE THREAD

**Issue by laurentpayot** [sveltejs/sapper#368]
I'm evaluating Sapper/Svelte, and it's been great so far :+1: 

However setting Firebase hosting was tricky because of weird Sapper behavior:

- [serving the firebase function](https://firebase.google.com/docs/functions/http-events#using_existing_express_apps) of the build wasn't working…
```
Error: ENOENT: no such file or directory, open '/home/laurent/projects/myproject/firebase/functions/.sapper/dev/client_assets.json'
    at Object.openSync (fs.js:434:3)
    at Object.readFileSync (fs.js:339:35)
    at fs.readFileSync (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:286:22)
    at handle_page (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:306:53)
    at find_route (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:540:6)
    at go (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:557:5)
    at /home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:559:6
    at find_route (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:275:3)
    at go (/home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:557:5)
    at /home/laurent/projects/myproject/firebase/functions/node_modules/sapper/src/middleware.ts:559:6
```
…untill setting the environment variable `SAPPER_DEST` to the build path.

- I first used `node firebase/functions/ssrapp` to have the build folder inside my functions folder but I had an error…
```
ENOENT: no such file or directory, open '/home/laurent/projects/myproject/firebase/functions/app/template.html'
```
…untill I changed it to `node firebase/functions/app` and made changes accordingly. Is there an `app` folder hardcoded somewhere? 

To sum it up here is the _simplified_ (I run "normal" Polka server if node env is 'development' to use `npm run dev`) code I use to host my Sapper app on Firebase:
```js
// app/server.js
import * as functions from 'firebase-functions'
import * as express from 'express'
import sapper from 'sapper'
import { manifest } from './manifest/server.js'

process.env.SAPPER_DEST = 'app' // needed

const app = express() // Polka is not working with Firebase
  .use(
    // compression is done by Firebase Nginx servers
    // assets folder static files are served by Firebase hosting
    sapper({ manifest })
  )

export default functions.https.onRequest(app)
```

Hope it helps Firebase users.

**Comment by lukeed** [comment]
:wave: Do you know what was causing Polka to not work? Or at least what kind of erroiyou were getting? Thanks !

**Comment by laurentpayot** [comment]
with
```js
const app = polka().use(sapper({ manifest }))
```
I get:

{"stack":"TypeError: handler is not a function\n    at cloudFunction (/home/laurent/projects/myproject/firebase/functions/node_modules/firebase-functions/lib/providers/https.js:57:9)\n    at app.use (/usr/local/lib/node_modules/firebase-tools/node_modules/@google-cloud/functions-emulator/src/supervisor/worker.js:151:11)\n    at Layer.handle [as handle_request] (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:317:13)\n    at /usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:335:12)\n    at next (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:275:10)\n    at app.use (/usr/local/lib/node_modules/firebase-tools/node_modules/@google-cloud/functions-emulator/src/supervisor/worker.js:123:7)\n    at Layer.handle [as handle_request] (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:317:13)","message":"handler is not a function","name":"TypeError"}

Maybe because Firebase uses Express, at least for handling requests without a server instance created: https://firebase.google.com/docs/hosting/functions#create_an_http_function_to_your_hosting_site

**Comment by laurentpayot** [comment]
With Polka and
```js
export default functions.https.onRequest(app.handler)
```
I get:

{"stack":"TypeError: Cannot set property path of #<IncomingMessage> which has only a getter\n    at Polka.handler (/home/laurent/projects/myproject/firebase/functions/node_modules/polka/index.js:73:12)\n    at cloudFunction (/home/laurent/projects/myproject/firebase/functions/node_modules/firebase-functions/lib/providers/https.js:57:9)\n    at app.use (/usr/local/lib/node_modules/firebase-tools/node_modules/@google-cloud/functions-emulator/src/supervisor/worker.js:151:11)\n    at Layer.handle [as handle_request] (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/layer.js:95:5)\n    at trim_prefix (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:317:13)\n    at /usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:284:7\n    at Function.process_params (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:335:12)\n    at next (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/index.js:275:10)\n    at app.use (/usr/local/lib/node_modules/firebase-tools/node_modules/@google-cloud/functions-emulator/src/supervisor/worker.js:123:7)\n    at Layer.handle [as handle_request] (/usr/local/lib/node_modules/firebase-tools/node_modules/express/lib/router/layer.js:95:5)","message":"Cannot set property path of #<IncomingMessage> which has only a getter","name":"TypeError"}

**Comment by lukeed** [comment]
Interesting, thanks~! I know that basic Sapper + Polka apps work fine on Firebase hosting, but I'll have to dig into more complex use cases to see where it's falling apart. 

If you have an idea of what might be throwing this error, it'd be appreciated. Thanks!

**Comment by laurentpayot** [comment]
@lukeed I will try some changes and I will let you know if there is something.

Are the Sapper + Polka + Firebase hosting apps you know on github? I'm curious to see how they did it…

**Comment by lukeed** [comment]
This one was shared in the Discord circa last week: https://github.com/fusionstrings/firebase-functions-sapper

[Live Demo](https://project-fusion-sapper.firebaseapp.com/)

**Comment by laurentpayot** [comment]
Thanks! It looks like they put a Sapper project inside a Firebase functions directory. It's a bit counter intuitive. I followed the way recommended by Firebase for Nuxt.js, i.e. putting a functions directory inside a Nuxt/Sapper project: https://youtu.be/ZYUWsjUxxUQ?t=106

**Comment by kylecordes** [comment]
Right, it is possible to adjust the settings used for firebase functions, to avoid having to move all of the files "down" a directory.

**Comment by laurentpayot** [comment]
@lukeed Maybe I'm wrong but [in this demo project](https://github.com/fusionstrings/firebase-functions-sapper/blob/fe7e3bfb574251efe45e4f0171cd09bedd90fffc/functions/index.js#L7) don't they simply use [request and response objects](https://firebase.google.com/docs/functions/http-events#using_express_request_and_response_objects) instead of using an [existing instance of a server](https://firebase.google.com/docs/functions/http-events#using_existing_express_apps)?
If I'm right then in this example Polka isn't really used by Google Cloud functions to serve content…

**Comment by lukeed** [comment]
I didn't check it out closely tbh. 

I think something you can try @laurentpayot is this:

```js
let { handler } = app;
export default functions.https.onRequest((req, res) => {
  handler(req, res);
})
```

My guess is that the `functions` utility is passing in a third argument (eg, `next`). Polka _takes_ a third argument, but expects an object instead (via Next.js, for pre-parsed `url` info).

That sounds like it fits the error message you posted.

Lemme know~!

**Comment by laurentpayot** [comment]
Thanks @lukeed yes it works, just like in the repo you mentioned.
You gave me the following idea:
```js
export default functions.https.onRequest(sapper({ manifest }))
```
is working because no app is needed within Firebase functions, but simply the request handler (sapper({ manifest })) :wink:

**Comment by lukeed** [comment]
Ah nice, bingo!

**Comment by rof20004** [comment]
My server.js is like this:
```
import sirv from 'sirv';
import polka from 'polka';
import compression from 'compression';
import * as sapper from '@sapper/server';
import bodyParser from 'body-parser';
import session from 'express-session';
import sessionFileStore from 'session-file-store';
import 'dotenv/config';

const { PORT, NODE_ENV, NOW, SECRET } = process.env;
const dev = NODE_ENV === 'development';
const FileStore = sessionFileStore(session);

function logger(req, res, next) {
    console.log(`~> Received ${req.method} on ${req.url}`);
    next();
}

function protect(req, res, next) {
	const allowed = [ '/credentials' ];
	const isPrivate = !allowed.includes(req.path);
	const isFile = req.path.includes('.');
	const hasToken = req.session.token;

	if (isPrivate && !isFile && !hasToken) {
		res.statusCode = 302;
		res.setHeader('Location', '/credentials');
		res.end();
	}

	next();
}

polka() // You can also use Express
	.use(bodyParser.json())
	.use(session({
		secret: SECRET,
		resave: false,
		saveUninitialized: false,
		cookie: { maxAge: 31536000 },
		store: new FileStore({ path: NOW ? `/tmp/sessions` : `.sessions` })
	}))
	.use(protect)
	.use(
		compression({ threshold: 0 }),
                sirv('static', { dev }),
		sapper.middleware({
			session: req => ({ user: req.session && req.session.user })
		})
	)
	.listen(PORT, err => {
		if (err) console.log('error', err);
	});
```

My firebase functions/index.js is like this:
```
const functions = require('firebase-functions');
const express = require('express');
const { sapper } = require('./__sapper__/build/server/server');
const app = express().use(sapper.middleware());

exports.ssr = functions.https.onRequest(app);
```

Someone can help me? I got this error: TypeError: Cannot read property 'middleware' of undefined

**Comment by rof20004** [comment]
Hi all,

I solved these two errors:

1. My SECRET variable was empty.
2. I removed public from firebase.json, I do not need it, I think.

But now when I enter in /home it not redirects to /credentials, some help please '-'.
??

**Comment by atomobianco** [comment]
This repository by @sanderhahn resumes the original way to deploy to Cloud Run and adds custom tips for the special case of Firebase
https://github.com/sanderhahn/gcloud-sapper

**Comment by kaleidawave** [comment]
After running `npm run build` and copying `__sapper__` into the `functions` folder I am still having problems

```js
const functions = require('firebase-functions');
const express = require('express');
const { sapper } = require('./__sapper__/build/server/server');

const app = express().use(sapper.middleware());
exports.app = functions.https.onRequest(app);
```

Running `firebase serve` returns:

```
TypeError: Cannot read property 'middleware' of undefined
```

**Comment by evdama** [comment]
go out one level i.e.
```
const { sapper } = require('../__sapper__/build/server/server');
```

**Comment by kaleidawave** [comment]
@evdama What? I have my `__sapper__` in the functions folder so that will not fix it...?

**Comment by evdama** [comment]
you're calling `__sapper__` from `lib/index.js` no? if so you must leave `lib/` first, then into `__sapper__/`

**Comment by kaleidawave** [comment]
No, that's only for typescript functions...

**Comment by kaleidawave** [comment]
nm, turns out I needed to append `export { sapper }` at the end of my `server.js` 😫

**Comment by evdama** [comment]
yes, so the ssr cloud function can use it

**Comment by pngwn** [comment]
This seems both resolved and a support issue.
